File: lib/commands/sunglasses.dart
import 'package:nomac/models/sunglasses/sunglasses.dart';
import 'package:nomac/util/constants.dart';
import 'package:nyxx/nyxx.dart';
import 'package:nyxx_interactions/interactions.dart';

import '../models/slash_command.dart';
import '../service_locator.dart';
import '../services/sunglasses_service.dart';
import '../util/is_admin.dart';

class SunglassesCommand extends NomacSlashCommand {
  final _sunglassesService = di<SunglassesService>();

  SunglassesCommand() : super('Sunglasses 😎');

  @override
  SlashCommandBuilder build() {
    return SlashCommandBuilder(
      'sunglasses',
      '😎',
      [
        CommandOptionBuilder(
          CommandOptionType.subCommand,
          'own',
          'Le epically own someone',
          options: [
            CommandOptionBuilder(
              CommandOptionType.user,
              'user',
              'Who you want to epically own',
              required: true,
            ),
          ],
        )..registerHandler(_own),
        CommandOptionBuilder(
          CommandOptionType.subCommand,
          'count',
          'Check how many times someone has been le epically owned by a mod',
          options: [
            CommandOptionBuilder(
              CommandOptionType.user,
              'user',
              'Someone other than yourself',
            ),
          ],
        )..registerHandler(_count),
        CommandOptionBuilder(
          CommandOptionType.subCommand,
          'ownage',
          'Check how many times someone has le epically owned someone else',
          options: [
            CommandOptionBuilder(
              CommandOptionType.user,
              'user',
              'Someone other than yourself',
            ),
          ],
        )..registerHandler(_ownage),
        CommandOptionBuilder(
          CommandOptionType.subCommand,
          'leaderboard',
          'Find out who has been epically owned the most',
        )..registerHandler(_leaderboard),
      ],
      guild: 285051699717210113.toSnowflake(),
    );
  }

  Future<void> _own(SlashCommandInteractionEvent event) async {
    if (!await isAdmin(event.interaction.memberAuthor.id, event.interaction.guild!.getFromCache()!)) {
      return event.respond(MessageBuilder.content('Nice try 😎'));
    }

    final String id = getArg(getSubCommand(event, 'own'), 'user')?.value;

    final user = await bot.fetchUser(id.toSnowflake());

    await _sunglassesService.own(Sunglasses(
      victim: id,
      ownedBy: event.interaction.memberAuthor.id.toString(),
      date: event.receivedAt,
    ));

    final count = await _sunglassesService.ownedCount(id);

    await event.respond(MessageBuilder.content('${user.mention} has now been 😎\'d **$count times!**'));
  }

  Future<void> _count(SlashCommandInteractionEvent event) async {
    final String id =
        getArg(getSubCommand(event, 'count'), 'user')?.value ?? event.interaction.memberAuthor.id.toString();

    final isMe = id == event.interaction.memberAuthor.id.toString();

    final count = await _sunglassesService.ownedCount(id);

    if (isMe) {
      return await event.respond(MessageBuilder.content('You have been 😎\'d **$count times!**'));
    } else {
      final username = (await bot.fetchUser(id.toSnowflake())).username;
      return await event.respond(MessageBuilder.content('$username has been 😎\'d **$count times!**'));
    }
  }

  Future<void> _ownage(SlashCommandInteractionEvent event) async {
    final String id =
        getArg(getSubCommand(event, 'ownage'), 'user')?.value ?? event.interaction.memberAuthor.id.toString();

    final isMe = id == event.interaction.memberAuthor.id.toString();

    final count = await _sunglassesService.ownageCount(event.interaction.memberAuthor.id.toString());

    if (isMe) {
      return await event.respond(MessageBuilder.content('You have le epically owned noobs **$count times!** 😎'));
    } else {
      final username = (await bot.fetchUser(id.toSnowflake())).username;
      return await event.respond(MessageBuilder.content('$username has le epically owned noobs **$count times!** 😎'));
    }
  }

  Future<void> _leaderboard(SlashCommandInteractionEvent event) async {
    await event.acknowledge();
    final leaderboard = await _sunglassesService.getTop();

    final embed = EmbedBuilder()
      ..author = embedAuthor
      ..color = nomacDiscordColor;

    for (var element in leaderboard) {
      User? user;

      try {
        user = await bot.fetchUser(element.discordId.toSnowflake());
      } catch (err) {
        user = null;
      }
      embed.addField(
        name: '#${leaderboard.indexOf(element) + 1}',
        content: '${user != null ? user.username : element.discordId} - ${element.count}',
      );
    }

    await event.respond(MessageBuilder.embed(embed));
  }
}

File: src/commands/sunglasses.ts
import { SlashCommandBuilder } from "@discordjs/builders";
import { CommandInteraction, MessageEmbed } from "discord.js";
import pino from "pino";
import { container } from "tsyringe";
import { client } from "..";
import SunglassesService from "../services/sunglasses_service";

export default new SlashCommandBuilder()
	.setName("sunglasses")
	.setDescription("😎")
	.addSubcommand((option) =>
		option
			.setName("own")
			.setDescription("Le epically own someone")
			.addUserOption((user) =>
				user.setName("user").setDescription("Who you want to epically own").setRequired(true)
			)
	)
	.addSubcommand((option) =>
		option
			.setName("count")
			.setDescription("Check how many times someone has been le epically owned by a mod")
			.addUserOption((user) => user.setName("user").setDescription("Someone other than yourself"))
	)
	.addSubcommand((option) =>
		option
			.setName("ownage")
			.setDescription("Check how many times someone has le epically owned someone else")
			.addUserOption((user) => user.setName("user").setDescription("Someone other than yourself"))
	)
	.addSubcommand((option) =>
		option.setName("leaderboard").setDescription("Find out who has been epically owned the most")
	);

export const sunglassesOwn = async (interaction: CommandInteraction) => {
	const sunglasses = container.resolve(SunglassesService);

	const user = interaction.options.getUser("user", true);

	// TODO: Fix permissions checking. This doesn't work
	// if (!((interaction.member!.permissions as string & 0x8) == 0x8)) {
	if (true) {
		return interaction.reply({ content: "Nice try 😎", ephemeral: true });
	}

	try {
		await sunglasses.own(user.id, interaction.user.id);
		const count = await sunglasses.ownedCount(user.id);

		return interaction.reply({
			content: `${user.username} has been le epically owned ${count} times`,
		});
	} catch (error) {
		return interaction.reply({
			content: `An unknown error occurred`,
			ephemeral: true,
		});
	}
};

export const sunglassesCount = async (interaction: CommandInteraction) => {
	const sunglasses = container.resolve(SunglassesService);

	const user = interaction.options.getUser("user") ?? interaction.user;

	try {
		const res = await sunglasses.ownedCount(user.id);

		return interaction.reply({
			content: `${user.username} has been le epically owned ${res} times`,
		});
	} catch (error) {
		return interaction.reply({
			content: `An unknown error occurred`,
			ephemeral: true,
		});
	}
};

export const sunglassesOwnage = async (interaction: CommandInteraction) => {
	const sunglasses = container.resolve(SunglassesService);

	const user = interaction.options.getUser("user") ?? interaction.user;

	try {
		const res = await sunglasses.ownageCount(user.id);

		return interaction.reply({
			content: `${user.username} has epically owned n00bs ${res} times`,
		});
	} catch (error) {
		return interaction.reply({
			content: `An unknown error occurred`,
			ephemeral: true,
		});
	}
};

export const sunglassesLeaderboard = async (interaction: CommandInteraction) => {
	const sunglasses = container.resolve(SunglassesService);

	try {
		const res = await sunglasses.getTop();

		const names = [];
		for await (const item of res) {
			const user = await client.users.fetch(item.discordId).catch(() => null);

			names.push(user?.username ?? "[Unknown User]");
		}

		const embed = new MessageEmbed()
			.setColor("#B972DA")
			.setTitle(`😎 Leaderboard`)
			.addFields(
				names.map((e, i) => ({
					name: `#${i + 1}`,
					value: `${e} - ${res[i].count}`,
				}))
			);

		return interaction.reply({ embeds: [embed] });
	} catch (error) {
		pino({ prettyPrint: true }).error(error as any);
		return interaction.reply({
			content: `An unknown error occurred`,
			ephemeral: true,
		});
	}
};
